nixos/modules/services/games/minecraft-server.nix

```diff
@@ -4,6 +4,68 @@ with lib;
 
 let
   cfg = config.services.minecraft-server;
+  eulaFile = pkgs.writeText "eula.txt" ''
+    # eula.txt manged by NixOS Configuration
+    eula=${if cfg.eula then "true" else "false"}
+  '';
+  whitelistFile = pkgs.writeText "whitelist.json" (builtins.toJSON (lib.mapAttrsToList (n: v: { uuid = n; name = v; }) (if cfg.whitelist != null then cfg.whitelist else {}) ));
+
+  # See https://minecraft.gamepedia.com/Server.properties
+  # for description of each property and allowed values
+  defaultServerProperties = {
+    allow-flight = false;
+    allow-nether = true;
+    difficulty = 1;
+    enable-command-block = false;
+    enable-query = false;
+    enable-rcon = false;
+    enforce-whitelist = false;
+    force-gamemode = false;
+    gamemode = 0;
+    generate-structures = true;
+    generator-settings = "";
+    hardcore = false;
+    level-name = "world";
+    level-seed = "";
+    level-type = "DEFAULT";
+    max-build-height = 256;
+    max-players = 20;
+    max-tick-time = 60000;
+    max-world-size = 29999984;
+    motd = "A NixOS Minecraft Server";
+    network-compression-threshold = 256;
+    online-mode = true;
+    op-permission-level = 4;
+    player-idle-timeout = 0;
+    prevent-proxy-connections = false;
+    pvp = true;
+    "query.port" = 25565;
+    "rcon.password" = "";
+    "rcon.port" = 25575;
+    resource-pack = "";
+    resource-pack-sha1 = "";
+    server-ip = "";
+    server-port = 25565;
+    snooper-enabled = false;
+    spawn-animals = true;
+    spawn-monsters = true;
+    spawn-npcs = true;
+    spawn-protection = 16;
+    use-native-transport = true;
+    view-distance = 10;
+    white-list = false;
+  };
```

Comment:
If you have the type of the `serverProperties` option specified correctly, you can just put these values into the `config` section like `minecraft-server.serverProperties = { ... };`. This will then merge it automatically into the user specified values. To have the user override them however you need a `mkDefault` for each one of the options though (use a `mapAttrs` for that), I think a single `serverProperties = mkDefault { ... }` would remove all defaults if the user set any value.

Reply:
Also: Is there a problem with just not setting these defaults? Can the minecraft server handle unset values and default them itself? Because that would be preferable, because if defaults change, we wouldn't have to update them here.

Reply:
Yes the minecraft server can handle missing values. I did this mostly for documentation on what all the correct key, values are. I agree that the default values may change over time.